src/index.js

```diff
@@ -287,16 +291,25 @@ class Bitswap {
 
         yield block
 
-        self.notifications.hasBlock(block)
-        self.engine.receivedBlocks([block.cid])
-        // Note: Don't wait for provide to finish before returning
-        self.network.provide(block.cid).catch((err) => {
-          self._log.error('Failed to provide: %s', err.message)
-        })
+        self._sendHaveBlockNotifications(block)
```

Comment:
It's my understanding that the async function `asyncFn` passed to `blockstore.putMany(asyncFn)` supplies the blocks to be put - but doesn't make guarantee about when those blocks will be commited to the blockstore.

So I think it may make more sense to send notifications after blockstore.putMany() has completed.

Reply:
I thought similar, but you'd have to iterate over the list of blocks twice and they may not be available to the source any more (e.g. it may not be an array).

Either that or you store the blocks in a list for processing after `putMany`, which may exhaust available memory, or you store just the CIDs and retrieve the blocks after the `putMany`, though they may not be available and the list may be unreasonably long, etc..

Reply:
The solution I am using in the new message types PR is to store them in an array:
https://github.com/ipfs/js-ipfs-bitswap/pull/211/files/f2b6be5ffbae25bdd911496a356b02b4490714ce#diff-1fdf421c05c1140f6d71444ea2b27638R285

I don't think it will use much more memory, because it's just a list of pointers, right?

Reply:
It's going to be a factor of the size of the added blocks - if you add a 10GB .iso file to IPFS, sticking them all in an array before sending notifications will cause the process to consume 10GB of memory.

If instead each block is processed one at a time (or in small batches) from reading the file to writing it out to disk it'll consume a lot less.

This isn't the only problem site, obviously - if our final blockstore batch size is unbounded it'll crash there too.

Reply:
Also, I’m not sure bitswap is the right place to solve this problem.  The blockstore has created it by making the behaviour of putMany non-obvious (to me at least), it should make the caller’s life a bit easier by limiting batch sizes, caching things that have been batched up, etc, though this is way out of scope for this PR.

Reply:
IIUC the blocks are a parameter to `bitswap.putMany(blocks)`, so they're not going to get cleaned up until the function returns. If we create an array of pointers to the blocks, that will use a negligible amount of memory compared to the blocks themselves.

I agree that the blockstore is probably a better place to tackle memory issues.

Just to clarify, when we call `yield` on line 292 I don't believe that means that the block has been committed to the blockstore. That's why I'm suggesting we only send out notifications after `blockstore.putMany()` completes.

Reply:
`bitswap.putMany(blocks)` takes an iterable as an argument, which might be an array in which case yes, they are all in memory at once so storing references to them in an array does not hurt, but it also might be a readable stream or a generator in which case they are pulled one block at a time.  It's for this case that we do not want to stick them all in an array.

Reply:
Ahh I see you're right `blocks` is an async iterable.

For correctness I believe we need to send notifications after the block has been committed, ie at the end of `blockstore.putMany()`.

In order to be able to send notifications we could check the type of `blocks` and then
- if it's an array: store a list of refs to each block that was put
- if it's an async iterable: call `datastore.getBlocks()` for notifications
  Probably most datastore implementations will have some caching that will make this perform reasonably well

Reply:
Looking back on this `blockstore.putMany()` conflates batching with streaming which I think is a mistake.

The intention (as I see it) of `putMany` is to allow streaming blocks into the blockstore, but it can actually end up batching those blocks, worst case storing them until the stream ends - and in our case forcing the caller to store something to do with the whole stream (e.g. the list of CIDs to notify of)

I think we should be explicit in our interfaces and remove the batching semantic from `blockstore.putMany()` and instead expose a `blockstore.putBatch()` or similar method for when that behaviour is desirable.

The batching is done [at the repo level](https://github.com/ipfs/js-ipfs-repo/blob/master/src/blockstore.js#L91-L105), I'm more than happy to pull that out.

Reply:
It's been pulled out. `ipfs-repo@3.x.x` `.putMany()` now streams to the datastore instead of using batches.